Detection with cpuid that would work for N2

paolotricerri opened the pull request:
This PR just aims at presenting an approach to use the cpuid field for aarch64 cores. 

This patch wants to address the question raised here: https://github.com/archspec/archspec-json/pull/105 about what to do when PR https://github.com/archspec/archspec-json/pull/106 and https://github.com/archspec/archspec-json/pull/105  will be merged and neoverse_v2 and neoverse_n2 would end up with the same list of features. So, how would we distinguish them as based only on features?

It introduces a new Microarchitecture, the AARCH64_Microarchitecture, which contains the cpuid field, if present, and then it uses it for compatibility checks. 

The changes on the test_cpu file is because in the archspec-json iI have introduced a data file for neoverse_n2 to test it.

CC: @dslarm, @alalazo

alalazo commented on archspec/cpu/microarchitecture.py:
```diff
@@ -300,6 +299,18 @@ def tuplify(ver):
         msg = msg.format(self.name, compiler, version)
         raise UnsupportedMicroarchitecture(msg)
 
+class AARCH64_Microarchitecture(Microarchitecture):
+    """Represents a specific AARCH64 CPU micro-architecture.
+
+    The only addition to the Microarchitecture class it inherits
+    from is the cpuid field that for specific aarch64 architectures
+    should be present in the definition of the micro-architecture.
+    """
+
+    def __init__(self, name, parents, vendor, features, compilers, generation=0, cpuid = None):
+        self.cpuid = cpuid
+        super().__init__(name, parents, vendor, features, compilers, generation)
+
```
Let's not create a separate class, and just add a new field to `Microarchitecture`. We already have `generation` which is relevant only on `ppc`.

paolotricerri replied:
Agreed and removed.

alalazo commented on .gitmodules:
```diff
@@ -1,6 +1,6 @@
 [submodule "archspec/json"]
 	path = archspec/json
-	url = https://github.com/archspec/archspec-json.git
+	url = https://github.com/paolotricerri/archspec-json.git
```
This, of course, needs to be undone before merging :slightly_smiling_face:

paolotricerri replied:
Of course :). Was just trying to make sure tests were running and passing. Happy to restore correct value as soon as the review about `cpuid` vs `cpupart` is closed.

paolotricerri commented:
Hi @alalazo, I just thought of explaining my changes as you will see I have kept some `cpuid` parts and changed to `cpupart` somewhere else.

I agree with you that the `cpuid` in the Json file contained information that were already contained in the conversion table at the bottom of the microarchitectures.json file and that we would only really need the `cpupart` which is what I have changed in the json file.

However, in the actual `detect.py` code, I have decided to keep the `cpuid` field and pass it to the constructor because i thought it was a way to check at the same time both the vendor (first part of cpuid) and the cpu part (second part) and so I thought it was a more effective way to check both fields for microarchitecture detection rather than having always to check the two fields separately. I hope this makes sense. Happy to reduce to only checking the cpupart if you think that is enough.

Thanks
P

paolotricerri commented:
HI @alalazo , apologies for keep pestering :)!

I was wondering if you had the time to look at the proposed changes after your comments? 

Thank you for your patience!

alalazo commented:
@paolotricerri Sorry for the delay on https://github.com/archspec/archspec/pull/170#issuecomment-2255424487 I just got back from holyday. Will have a look this week.

paolotricerri commented:
Hi @alalazo , thanks for your message. 

I just received the message of the failing workflows. From what I can see in the picture, I am wondering if py36 and py38 are the versions of python used for those workflows. If so, I'll try to install them and run the tests with that version of python to make the test successful for those versions of python too.

![image](https://github.com/user-attachments/assets/45b49af8-1cb3-42ed-b997-a1be35e207c7)

alalazo commented:
@paolotricerri I would proceed like this:
1. Finish https://github.com/archspec/archspec-json/pull/106 (basically add the missing `/proc/cpuinfo` sample) and merge it 
2. Point the submodule here to the head of archspec-json after we the PR at 1 is merged
3. Finalize this PR

Would that be fine with you?

alalazo commented on tests/test_cpu.py:
```diff
@@ -102,6 +102,39 @@ def _open(not_used_arg):
+        "linux-rhel8-neoverse_v1",
+    ]
+)
+def expected_linux_target(request, monkeypatch):
```
There is no need to add a cut and paste fixture, which mimics `expected target`. Let's just add new tests to `expected_targets` and ensure everything is still detected correctly.

paolotricerri replied:
Hi @alalazo  I have addressed this point but I'll move over to https://github.com/archspec/archspec/pull/177 and push it there.

alalazo commented on tests/test_cpu.py:
```diff
@@ -201,6 +234,14 @@ def test_target_detection(expected_target):
+    detected_target = archspec.cpu.host()
+    if expected_linux_target == 'thunderx2':
+        ## We don't have cpuid for thunderx2 yet
+        assert detected_target.name == "", "Unexpected non-empy target for thunderx2"
```
:thinking: This would be a regression The cpupart for `thunderx2` is `0x0af` (it's in the sample used for tests).

paolotricerri replied:
Would you just use the cpupart then? I think it is fine for us too.

alalazo replied:
@paolotricerri Yes, see #177 as a tentative implementation.

paolotricerri replied:
I'll ask for @dslarm's point of view and get back to you on that after our chat next Monday. Thanks

alalazo commented:
@paolotricerri Would you mind if I push to this branch?

alalazo commented on archspec/cpu/microarchitecture.py:
```diff
@@ -73,13 +73,14 @@ class Microarchitecture:
     feature_aliases = FEATURE_ALIASES
 
-    def __init__(self, name, parents, vendor, features, compilers, generation=0):
+    def __init__(self, name, parents, vendor, features, compilers, generation=0, cpuid=""):
```
We should use `cpupart` everywhere, including when we read the JSON file. There is no `cpuid` filed in the JSON...

paolotricerri replied:
Hi @alalazo , I agree that having cpupart in the Json and the cpuid here can be confusing. However, as I explained above, @dslarm and I would prefer, at this stage, to add the necessary fields only for the ARM vendor (CPU Implementer 0x41) and not have it for all ARM based uarchs. threrefore I thought using cpuid anyway was going to enforce the use of CPU Implementer. Can change and use just the cpupart.

paolotricerri commented:
> @paolotricerri I would proceed like this:
> 
>     1. Finish [Neoverse V2: remove `sm3`, `sm4` and `svesm4` archspec-json#106](https://github.com/archspec/archspec-json/pull/106) (basically add the missing `/proc/cpuinfo` sample) and merge it
> 
>     2. Point the submodule here to the head of archspec-json after the PR at 1 is merged
> 
>     3. Finalize this PR
> 
> 
> Would that be fine with you?

Yes, I was trying to untangle a way to proceed but this sounds good. Thanks. 

I have booked some time with @dslarm to go over the questions for neoverse_v* next Monday. Will get back to that PR and this after that. Thanks

paolotricerri commented:
> @paolotricerri Would you mind if I push to this branch?

No. Do you mean to the main archspec repo and we continue development there? Anything is good with me. Just so I know how to update my remotes and refs :)

alalazo commented:
@paolotricerri Apologies for the confusion, the most up-to-date PR is #177, if that works for you too.
